Add support for alternative CSV format

markus456 opened the pull request:
Adds rudimentary CSV parsing of payment records in the form of the parseCSV function. Also added unit tests with the currently known format of the CSV schema. Changes to the format are somewhat easy to do since it uses DictReader.

Added a new button to the transactions admin page for uploading the new CSV files. The old button for the NDA forms is left there for backwards compatibility.

Fixed the parser.py unit test to use the python from the environment instead of a hard-coded path. This makes it work with venv.

markus456 committed:
Add support for alternative CSV format

Adds rudimentary CSV parsing of payment records in the form of the
parseCSV function. Also added unit tests with the currently known format
of the CSV schema. Changes to the format are somewhat easy to do since
it uses DictReader.

Added a new button to the transactions admin page for uploading the new
CSV files. The old button for the NDA forms is left there for backwards
compatibility.

Fixed the parser.py unit test to use the python from the environment
instead of a hard-coded path. This makes it work with venv.

markus456 commented:
@rambo if possible, please review this change. I suspect that my Django usage is not quite appropriate.

rambo reviewed (approved):
it's been ages since I touched Django but it doesn't look any worse than the original to current me(These days I'd expect any new code to pass `mypy --strict`)

rambo commented on project/ndaparser/parser.py:
```diff
@@ -126,6 +127,22 @@ def ascii2scandic(string):
+            continue
+        amount = int(float(t["Määrä"].replace(",", ".")))
+        timestamp = datetime.strptime(t["Kirjauspäivä"], "%Y/%m/%d").date()
+        archive_id = "TODO: We need the archive ID"
```
Change this to something that makes for example sha1 from the information of the transaction. Actual Nordea transaction-ids will not be available.

markus456 replied:
I did as you suggested and went with the SHA1 of the CSV record fields concatenated into a single string.

markus456 committed:
Use SHA1 of the CSV record as archive ID

This is an adequate substitute of the archive ID from the NDA records.

markus456 committed:
Make the test cases better

Repeatedly testing the same thing is somewhat pointless. Testing that
certain missing fields cause the record to be rejected is better. Also
added assertions so that the result is actually verified.

Changed the parsing to only accept rows with a valid amount and archive
date. Without these the transaction object would not work. A missing
name is accepted as the reference number should be enough to uniquely
identify the member who paid it. A missing reference number is also
accepted in the hopes that the name can be used to figure out who the
real owner of the transaction is.

markus456 committed:
Add test case for negative amounts

Just to sanity check that the parser works with it. These will get
rejected by the transaction handler.

jssmk commented:
Good job updating the test file, starts to look good to me
